pkgs/development/libraries/hspell/default.nix

```diff
@@ -1,27 +1,62 @@
-{ stdenv, fetchurl, perl, zlib }:
+{ stdenv, fetchurl, buildPackages, zlib }:
 
-stdenv.mkDerivation rec {
-  name = "${passthru.pname}-${passthru.version}";
+# hspell uses a couple of binaries it generates during it's own build.
+# These will cause the build to fail later in cross-platform builds as
+# these binaries have the wrong executable format.
+# Using a hspell to bootstrap it's own build natively resolves this issue.
 
-  passthru = {
-    pname = "hspell";
-    version = "1.1";
-  };
+let
+  upstream_name = "hspell";
+  mkDrv = {
+    pname ? upstream_name,
+    version ? "1.1",
+    sha256 ? "08x7rigq5pa1pfpl30qp353hbdkpadr1zc49slpczhsn0sg36pd6",
+  }: rec {
 
-  PERL_USE_UNSAFE_INC = "1";
+    passthru = {
+      pname = pname;
+      version = version;
+    };
 
-  src = fetchurl {
-    url = "${meta.homepage}${name}.tar.gz";
-    sha256 = "08x7rigq5pa1pfpl30qp353hbdkpadr1zc49slpczhsn0sg36pd6";
-  };
+    patchPhase = ''patchShebangs .'';
+    preConfigure = stdenv.lib.optionalString (stdenv.hostPlatform != stdenv.buildPlatform) ''
+      substituteInPlace Makefile.in --replace "ar cr" "${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ar cr"
+      substituteInPlace Makefile.in --replace "ranlib" "${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}ranlib"
+      substituteInPlace Makefile.in --replace "STRIP=strip" "STRIP=${stdenv.lib.getBin stdenv.cc.bintools.bintools}/bin/${stdenv.cc.targetPrefix}strip"
+    '';
+
+    PERL_USE_UNSAFE_INC = "1";
 
-  patchPhase = ''patchShebangs .'';
-  buildInputs = [ perl zlib ];
+    name = "${upstream_name}-${version}";
+    src = fetchurl {
+      url = "${meta.homepage}${name}.tar.gz";
+      sha256 = sha256;
+    };
 
-  meta = with stdenv.lib; {
-    description = "Hebrew spell checker";
-    homepage = "http://hspell.ivrix.org.il/";
-    platforms = platforms.all;
-    license = licenses.gpl2;
+    nativeBuildInputs = [ buildPackages.perl ];
+    buildInputs = [ zlib ];
+
+    meta = with stdenv.lib; {
+      description = "Hebrew spell checker";
+      homepage = "http://hspell.ivrix.org.il/";
+      platforms = platforms.all;
+      license = licenses.gpl2;
+      maintainers = [ maintainers.kampka ];
+    };
   };
-}
+
+  bootstrap = stdenv.mkDerivation (mkDrv{ pname = "hspell-bootstrap"; } // {
```

Comment:
This shouldn't need to be in a separate derivation. I think you can just do something like:

```
preBuild = ''
  make CC=${buildPackages.stdenv.cc}/bin/cc find_sizes
'';
```

to avoid it. Ideally we should patch hspell to do the right thing (use $(CC) for hspell and $(BUILD_CC) for find_sizes).

Reply:
But building a bootstrap like this is not idiomatic, since lots of other packages have a find_sizes that they get correctly.

Reply:
> Ideally we should patch hspell to do the right thing (use $(CC) for hspell and $(BUILD_CC) for find_sizes).

Agreed this should be fixed upstream, but as it does not seem to be an actively maintained project, I'd rather have a working version here in any case.

> This shouldn't need to be in a separate derivation.

Done.